daemon/src/db/failed.rs

```diff
@@ -0,0 +1,179 @@
+//! This module allows us to move faile CFDs to a separate table so
+//! that we can reason about loading them independently from open
+//! CFDs.
+//!
+//! Therefore, it also provides an interface to load failed CFDs: the
+//! `FailedCfdAggregate` trait. Implementers of the trait will be able
+//! to call the `crate::db::load_all_cfds` API, which loads all types
+//! of CFD.
+
+use crate::db;
+use crate::db::delete_from_cfds_table;
+use crate::db::delete_from_events_table;
+use crate::db::load_cfd_row;
+use crate::db::CfdAggregate;
+use crate::db::Connection;
+use anyhow::Result;
+use model::Contracts;
+use model::Identity;
+use model::Leverage;
+use model::OrderId;
+use model::Position;
+use model::Price;
+use model::Role;
+use sqlx::Connection as _;
+use sqlx::Sqlite;
+use sqlx::Transaction;
+
+/// A trait for building an aggregate based on a `FailedCfd`.
+pub trait FailedCfdAggregate: CfdAggregate {
+    fn new_failed(args: Self::CtorArgs, cfd: FailedCfd) -> Self;
+}
+
+/// Data loaded from the database about a failed CFD.
+#[derive(Debug, Clone, Copy)]
+pub struct FailedCfd {
+    pub id: OrderId,
+    pub position: Position,
+    pub initial_price: Price,
+    pub taker_leverage: Leverage,
+    pub n_contracts: Contracts,
+    pub counterparty_network_identity: Identity,
+    pub role: Role,
+}
+
+impl Connection {
+    pub async fn move_to_failed_table(&self) -> Result<()> {
+        let ids = self.failed_cfds().await?;
+
+        if !ids.is_empty() {
+            tracing::debug!("Moving CFDs to failed_cfds table: {ids:?}");
+        }
+
+        for id in ids.into_iter() {
+            let pool = self.inner.clone();
+
+            let fut = async move {
+                let mut conn = pool.acquire().await?;
+                let mut db_tx = conn.begin().await?;
+
+                let cfd = load_cfd_row(&mut db_tx, id).await?;
+                insert_failed_cfd(&mut db_tx, cfd).await?;
+
+                delete_from_events_table(&mut db_tx, id).await?;
+                delete_from_cfds_table(&mut db_tx, id).await?;
+
+                db_tx.commit().await?;
+
+                anyhow::Ok(())
+            };
+
+            if let Err(e) = fut.await {
+                tracing::warn!(order_id = %id, "Failed to move failed CFD: {e:#}");
+            }
+
+            tracing::debug!(order_id = %id, "Moved CFD to `failed_cfds` table");
```

Comment:
You will print this even in an error case :)

The entire construct should probably be a `match`!